### ISSUE TITLE
[Experimental/Discussion]: Prompt auto-tuning in eval outer loop

### ISSUE THREAD

**Issue by jon-chuang** [run-llama/llama_index#6713]
### Feature Description

As part of hyperparameter optimization, rewrite prompts using an LLM predictor against a validation set. 

It is well known that small tweaks to prompts can lead to better adherence to desired behaviors.

**Grid Search:** 
The simple form of this is generate and grid search (no feedback).

**Evolutionary search:** 
Alternately, by feeding back the negative, neutral and positive examples of prompt rewriting (neutral - no statistically significant change), we can redirect the prompt semantically, and force the LLM to infer features in the prompts that are leading to more positive results, thus, it is hoped, produce better prompts. In this scenario, we stop the tuning process once the prompts generated show no more improvement over the baseline prompt (we have hit a local optimum). 

Nevertheless, one should be cautious not to overfit on the validation set.

A good pattern might be to optimize over a single prompt or a small handful of interrelated prompts at a time.

Edit:
Some prior work here https://github.com/mshumer/gpt-prompt-engineer
